src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -445,51 +492,154 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
                 {
-                    // consumes data ptr and size, pushes the instance
+                    // Consumes data ptr and size, pushes the instance.
                     _builder.EmitOpCode(ILOpCode.Newobj, stackAdjustment: -1);
                 }
 
-                EmitSymbolToken(ctor.AsMember(spanType), wrappedExpression.Syntax, optArgList: null);
+                EmitSymbolToken(rosPointerCtor.AsMember(spanType), wrappedExpression.Syntax, optArgList: null);
+                return true;
             }
 
+            // We're dealing with a primitive that's larger than a single byte.
+            Debug.Assert(specialElementType.SizeInBytes() is 2 or 4 or 8, "Supported primitives are expected to be 2, 4, or 8 bytes");
+
+            if (lengthForConstructor != elementCount)
+            {
+                // We need to use RuntimeHelpers.CreateSpan / cached array, but the code has requested a subset of the elements.
+                // That means the code is something like `new ReadOnlySpan<char>(new[] { 'a', 'b', 'c' }, 1, 2)`
+                // rather than `new ReadOnlySpan<char>(new[] { 'b', 'c' })`.  If such a pattern is found to be
+                // common, this could be augmented to accomodate it.  For now, we just return false to fail
+                // to optimize this case.
+                return false;
+            }
+
+            if (inPlaceTarget is not null)
+            {
+                // We can use RuntimeHelpers.CreateSpan / cached array but not for in-place initialization. Fail to optimize,
+                // but tell the caller they can call this again with a null inPlaceTarget, at which point this
+                // should be able to optimize the call.
+                avoidInPlace = true;
+                return false;
+            }
+
+            // As we're dealing with multi-byte types, endianness needs to be considered. Such handling is provided by the
+            // runtime's RuntimeHelpers.CreateSpan, which will wrap a span around the blob on little endian and which will
+            // allocate an array and cache it on big endian.
+            MethodSymbol? createSpan = (MethodSymbol?)_module.Compilation.GetWellKnownTypeMember(WellKnownMember.System_Runtime_CompilerServices_RuntimeHelpers__CreateSpanRuntimeFieldHandle);
+            if (createSpan is not null)
+            {
+                if (createSpan.GetUseSiteInfo().DiagnosticInfo?.Severity == DiagnosticSeverity.Error)
+                {
+                    // Something went wrong.
+                    return false;
+                }
+
+                // CreateSpan was available. Use it.
+                // ldtoken <PrivateImplementationDetails>...
+                // call ReadOnlySpan<elementType> RuntimeHelpers::CreateSpan<elementType>(fldHandle)
+                var field = _builder.module.GetFieldForData(data, alignment: (ushort)specialElementType.SizeInBytes(), wrappedExpression.Syntax, _diagnostics.DiagnosticBag!);
```

Comment:
>_diagnostics.DiagnosticBag! [](http://example.com/codeflow?start=145&length=27)

The guideline on nullability suppressions is to only use `!` if the method already checks, but the compiler wasn't able to analyze/track the information. Otherwise, an assertion is preferred.
Comment applies to multiple locations in this method.